be/src/util/frame_of_reference_coding.cpp

```diff
@@ -0,0 +1,377 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "util/frame_of_reference_coding.h"
+
+#include <algorithm>
+#include <cstring>
+
+#include "util/bit_util.h"
+#include "util/coding.h"
+
+namespace doris {
+
+static inline uint8_t bits(const uint64_t v) {
+    return v == 0 ? 0 : 64 - __builtin_clzll(v);
+}
+
+template<typename T>
+const T* ForEncoder<T>::copy_value(const T* in_data, size_t count) {
+    memcpy(&_buffered_values[_buffered_values_num], in_data, count * sizeof(T));
+    _buffered_values_num += count;
+    in_data += count;
```

Comment:
> Since the maximum bit width is 64, which only needs 6 bits to represent, I think we can use the first two bits (most significant) of BidWidth to represent OrderFlag, that is one byte metadata for each frame.

OK